### DISCUSSION TITLE
Reading the Real-Time Position after Simulation Completion in SOFA

### DISCUSSION THREAD

**Discussion by MONArfi** [SofaDefrost/SoftRobots#220] [Help - Programming with SoftRobots]
Hello all,

I am working with a soft trunk-like robot that has four cable actuators in SOFA, and I want to read its real-time position after the simulation is complete. In Python, I have created a MechanicalObject named 'dofs' which contains position and velocity. During the simulation, I can observe the position and velocity of 'dofs' under each cable. 

![RealTime_pos](https://github.com/SofaDefrost/SoftRobots/assets/92113874/9317a808-70a5-4fc6-8c9e-59db21fd2dea)

I have researched and found that I need to use WriteStates, open(), and read() functions to achieve this. However, I am unsure of the exact steps to do so.

In Python I have:
`cable_pos=cableL.addObject('MechanicalObject', name='dofs',
                             position=pullPoint[i]+[pos.toList() for pos in position])
`
1)	How should I modify the following line?

`cableL.createObject('WriteState', filename='cablePos.txt', writeF='true', period='0.1', listening="1")`

And, after the below code, 
`simulation = node.addChild('Simulation')`

2)	What should I write to read the position states?

**Comment by EulalieCoevoet** [comment]
Hello, 

This `cableL.createObject('WriteState', filename='cablePos.txt', writeF='true', period='0.1', listening="1")` will save the position of the cable's passing points. Is that what you need? 
Just in case, if you want extract the position of all the nodes of your FEM mesh, or just the end effector, you should put the component `WriteState` below the corresponding `MechanicalObject`.

Do you need to read those positions to analyse them? With another software maybe?

**Reply by MONArfi** [reply]
Hi,
Thanks for your response,
The above code results as follows:
 
![WriteState](https://github.com/SofaDefrost/SoftRobots/assets/92113874/f1ffbd4b-f205-4889-8284-4760f2016764)

Now, I just want to read it and simply draw that real-time data by Matplotlib. How can I read them in Python after the simulation completion?

Thank you

**Reply by EulalieCoevoet** [reply]
Well at this point you no longer needs SOFA. 
So I'll let you google how to read a file with python and how to use Matplotlib.

**Comment by MONArfi** [comment]
OK, Thank you very much.
